crates/energy/src/graph.rs

```diff
@@ -0,0 +1,242 @@
+#[derive(Debug, Clone)]
+pub enum Nearby {
+    Receiver(SmallVec<[Entity; 40]>),
+    Producer(SmallVec<[Entity; 40]>),
+}
+
+impl Nearby {
+    fn into_inner(self) -> SmallVec<[Entity; 40]> {
+        match self {
+            Nearby::Receiver(inner) => inner,
+            Nearby::Producer(inner) => inner,
+        }
+    }
+}
+
+#[derive(Component, Default, Debug, Clone)]
+pub struct NearbyUnits(SmallVec<[Nearby; 2]>, Option<Vec2>);
+
+fn setup(mut commands: Commands) {
+    commands.insert_resource(PowerGrid::default());
+}
+
+fn clean_up(mut commands: Commands) {
+    commands.remove_resource::<PowerGrid>();
+}
+
+/// This system spawns Energy Producers and Energy Receivers and nearby units.
+fn spawn_graph_components(
+    mut commands: Commands,
+    newly_spawned_units: Query<Entity, Added<Active>>,
+) {
+    for entity in newly_spawned_units.iter() {
+        commands
+            .entity(entity)
+            .insert((EnergyReceiver, NearbyUnits::default()));
+    }
+}
+
+fn update_nearby_recv(
+    spacial_index_producer: SpatialQuery<Entity, With<EnergyProducer>>,
+    spacial_index_receiver: SpatialQuery<Entity, With<EnergyReceiver>>,
+    mut nearby_units: Query<(Entity, &mut NearbyUnits, &Transform)>,
+) {
+    let time = Instant::now();
+    nearby_units
+        .par_iter_mut()
+        .for_each_mut(|(entity, mut nearby_units, transform)| {
+            if let Some(last_pos) = nearby_units.1 {
+                if transform.translation.to_flat().distance_squared(last_pos) < 0.5 {
+                    return;
+                }
+            }
+            nearby_units.1 = Some(transform.translation.to_flat());
+
+            let aabb = &Aabb::new(
+                Point::from(*(transform.translation - Vec3::splat(MAX_DISTANCE)).as_ref()),
+                Point::from(*(transform.translation + Vec3::splat(MAX_DISTANCE)).as_ref()),
+            );
+
+            let producers = spacial_index_producer.query_aabb(aabb, Some(entity));
+
+            let receivers = spacial_index_receiver.query_aabb(aabb, Some(entity));
+
+            update_nearby(nearby_units, producers.collect(), receivers.collect());
+        });
+    println!("update_nearby_recv: {:?}", time.elapsed());
+}
+
+fn update_nearby(
+    mut nearby_units: Mut<NearbyUnits>,
+    producers: Vec<Entity>,
+    receivers: Vec<Entity>,
+) {
+    let mut nearby_producers = SmallVec::new();
+    let mut nearby_receivers = SmallVec::new();
+
+    for producer in producers {
+        nearby_producers.push(producer);
+    }
+
+    for receiver in receivers {
+        nearby_receivers.push(receiver);
+    }
+
+    nearby_units.0 = smallvec![
+        Nearby::Producer(nearby_producers),
+        Nearby::Receiver(nearby_receivers)
+    ];
+}
+
+#[allow(clippy::type_complexity)]
+fn update_graph(
+    mut power_grid: ResMut<PowerGrid>,
+    nearby_units: Query<(Entity, &NearbyUnits), Or<(Added<NearbyUnits>, Changed<NearbyUnits>)>>,
```

Comment:
[`Chanded<>`](bevy::ecs::prelude::Changed) matches even added entities, therefore the whole `Or<>` can be simplified and the Clippy allow removed.